### ISSUE TITLE
Incompatibility with Stripe API version 2022-11-15

### ISSUE THREAD

**Issue by maxlaf** [woocommerce/woocommerce-gateway-stripe#2488]
**Describe the bug**
When UPE is enabled and the latest version of Stripe's API (2022-11-15) is used, various issues and notices are caused, the severity of which depends on the payment method used.

**To Reproduce**
1. Set your Stripe account to use API version 2022-11-15
2. Enable the new checkout experience (UPE) and put Stripe in test mode.
3. Enabled multiple payment methods. I used CCs and a deferred method, Bancontact.
4. Place an order with:
* A 3DSv2 required CC
* A regular CC
* Bancontact
5. See varying errors depending on the payment method.

**3DSv2 Required Card**

This seemed to produce the most severe problems out of the 3 methods I tested with. The checkout flow doesn't present any errors. However, once the order is placed and you view it on the backend, no charge is linked and the order is set to On Hold. But, the payment in Stripe succeeded.

In the Stripe logs, this error was thrown: `Could not find order via charge ID: ch_3M9GazI21hhR58QI0NUM2UqF`

![8c8gUa.png](https://user-images.githubusercontent.com/56649567/204405043-888eb170-034a-47de-bddb-d8efff485472.png)

**Regular CC and Bancontact**

Both of these payment methods completed in Stripe and the order status was changed to processing. But multiple notices were thrown with WP_DEBUG enabled:

<details>

```
Notice: Undefined variable: result in /srv/users/usercf38ea2b/apps/usercf38ea2b/public/wp-content/plugins/woocommerce-gateway-stripe/includes/class-wc-stripe-order-handler.php on line 294

Notice: Trying to get property 'id' of non-object in /srv/users/usercf38ea2b/apps/usercf38ea2b/public/wp-content/plugins/woocommerce-gateway-stripe/includes/class-wc-stripe-order-handler.php on line 294

Notice: Undefined variable: result in /srv/users/usercf38ea2b/apps/usercf38ea2b/public/wp-content/plugins/woocommerce-gateway-stripe/includes/class-wc-stripe-order-handler.php on line 298

Notice: Trying to get property 'id' of non-object in /srv/users/usercf38ea2b/apps/usercf38ea2b/public/wp-content/plugins/woocommerce-gateway-stripe/includes/class-wc-stripe-order-handler.php on line 298

Notice: Undefined variable: result in /srv/users/usercf38ea2b/apps/usercf38ea2b/public/wp-content/plugins/woocommerce-gateway-stripe/includes/class-wc-stripe-order-handler.php on line 304

Notice: Trying to get property 'balance_transaction' of non-object in /srv/users/usercf38ea2b/apps/usercf38ea2b/public/wp-content/plugins/woocommerce-gateway-stripe/includes/class-wc-stripe-order-handler.php on line 304

Notice: Trying to get property 'id' of non-object in /srv/users/usercf38ea2b/apps/usercf38ea2b/public/wp-content/plugins/woocommerce-gateway-stripe/includes/class-wc-stripe-order-handler.php on line 304

Notice: Undefined variable: result in /srv/users/usercf38ea2b/apps/usercf38ea2b/public/wp-content/plugins/woocommerce-gateway-stripe/includes/class-wc-stripe-order-handler.php on line 308

Notice: Undefined variable: result in /srv/users/usercf38ea2b/apps/usercf38ea2b/public/wp-content/plugins/woocommerce-gateway-stripe/includes/class-wc-stripe-order-handler.php on line 309
```

</details>

A different error was thrown in the Stripe logs: `Unable to update fees/net meta for order: 31`.

And the order itself had no charge linked at the top and the first charge ID in the order notes was empty:

![ANcORi.png](https://user-images.githubusercontent.com/56649567/204405402-91434734-71a8-40f5-ad66-616c0bfea0c3.png)

A common symptom across different payment methods is an empty charge ID in the order notes, and this error in the order notes: `Stripe charge authorized (Charge ID: ). Process order to take payment, or cancel to remove the pre-authorization. Attempting to refund the order in part or in full will release the authorization and cancel the payment. Order status changed from Pending payment to On hold.`

**Expected behavior**
I expected UPE to be compatible with the latets verison of Stripe's API.

**Workaround**
~~Disabling UPE and/or switching to a Stripe API version before 2022-11-15 solves these issues.~~
Disabling UPE does not solve this issue. Stripe API versions can only be rolled back within 72 hours of upgrading.

**Environment (please complete the following information):**
 - WordPress Version: 6.1.0
 - WooCommerce Version: 7.1.0
 - Stripe Plugin Version: 7.0.1
 - Browser [e.g. chrome, safari] and Version
 - Any other plugins installed

**Additional context**
Identified in: 5741056-zen

SSR:
<details>

```
`
### WordPress Environment ###

WordPress address (URL): https://continental-galliform.jurassic.ninja
Site address (URL): https://continental-galliform.jurassic.ninja
WC Version: 7.1.0
REST API Version: ✔ 7.1.0
WC Blocks Version: ✔ 8.7.5
Action Scheduler Version: ✔ 3.4.0
Log Directory Writable: ✔
WP Version: 6.1.1
WP Multisite: –
WP Memory Limit: 256 MB
WP Debug Mode: ✔
WP Cron: ✔
Language: en_US
External object cache: –

### Server Environment ###

Server Info: Apache/2.4.54 (Unix) OpenSSL/1.0.2g
PHP Version: 7.4.33
PHP Post Max Size: 1 GB
PHP Time Limit: 30
PHP Max Input Vars: 5000
cURL Version: 7.47.0
OpenSSL/1.0.2g

SUHOSIN Installed: –
MySQL Version: 5.7.33-0ubuntu0.16.04.1-log
Max Upload Size: 512 MB
Default Timezone is UTC: ✔
fsockopen/cURL: ✔
SoapClient: ✔
DOMDocument: ✔
GZip: ✔
Multibyte String: ✔
Remote Post: ✔
Remote Get: ✔

### Database ###

WC Database Version: 7.1.0
WC Database Prefix: wp_
Total Database Size: 5.80MB
Database Data Size: 4.22MB
Database Index Size: 1.58MB
wp_woocommerce_sessions: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
wp_woocommerce_api_keys: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_woocommerce_attribute_taxonomies: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
wp_woocommerce_downloadable_product_permissions: Data: 0.02MB + Index: 0.06MB + Engine InnoDB
wp_woocommerce_order_items: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
wp_woocommerce_order_itemmeta: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_woocommerce_tax_rates: Data: 0.02MB + Index: 0.06MB + Engine InnoDB
wp_woocommerce_tax_rate_locations: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_woocommerce_shipping_zones: Data: 0.02MB + Index: 0.00MB + Engine InnoDB
wp_woocommerce_shipping_zone_locations: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_woocommerce_shipping_zone_methods: Data: 0.02MB + Index: 0.00MB + Engine InnoDB
wp_woocommerce_payment_tokens: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
wp_woocommerce_payment_tokenmeta: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_woocommerce_log: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
wp_actionscheduler_actions: Data: 0.02MB + Index: 0.11MB + Engine InnoDB
wp_actionscheduler_claims: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
wp_actionscheduler_groups: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
wp_actionscheduler_logs: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_commentmeta: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_comments: Data: 0.02MB + Index: 0.09MB + Engine InnoDB
wp_links: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
wp_options: Data: 3.17MB + Index: 0.06MB + Engine InnoDB
wp_postmeta: Data: 0.08MB + Index: 0.06MB + Engine InnoDB
wp_posts: Data: 0.05MB + Index: 0.06MB + Engine InnoDB
wp_termmeta: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_terms: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_term_relationships: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
wp_term_taxonomy: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_usermeta: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_users: Data: 0.02MB + Index: 0.05MB + Engine InnoDB
wp_wc_admin_notes: Data: 0.02MB + Index: 0.00MB + Engine InnoDB
wp_wc_admin_note_actions: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
wp_wc_bookings_availability: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
wp_wc_bookings_availabilitymeta: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_wc_booking_relationships: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_wc_category_lookup: Data: 0.02MB + Index: 0.00MB + Engine InnoDB
wp_wc_customer_lookup: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_wc_download_log: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_wc_order_coupon_lookup: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_wc_order_product_lookup: Data: 0.02MB + Index: 0.06MB + Engine InnoDB
wp_wc_order_stats: Data: 0.02MB + Index: 0.05MB + Engine InnoDB
wp_wc_order_tax_lookup: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_wc_product_attributes_lookup: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
wp_wc_product_download_directories: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
wp_wc_product_meta_lookup: Data: 0.02MB + Index: 0.09MB + Engine InnoDB
wp_wc_rate_limits: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
wp_wc_reserved_stock: Data: 0.02MB + Index: 0.00MB + Engine InnoDB
wp_wc_tax_rate_classes: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
wp_wc_webhooks: Data: 0.02MB + Index: 0.02MB + Engine InnoDB

### Post Type Counts ###

attachment: 1
bookable_person: 4
page: 7
post: 2
product: 5
revision: 1
shop_order: 11
wp_global_styles: 1

### Security ###

Secure connection (HTTPS): ✔
Hide errors from visitors: ❌Error messages should not be shown to visitors.

### Active Plugins (5) ###

Bookings Helper: by WooCommerce – 1.0.3
Companion Plugin: by Osk – 1.28
Jetpack: by Automattic – 11.5.1
WooCommerce Stripe Gateway: by WooCommerce – 7.0.1
WooCommerce: by Automattic – 7.1.0

### Inactive Plugins (4) ###

Akismet Anti-Spam: by Automattic – 5.0.1
Hello Dolly: by Matt Mullenweg – 1.7.2
WooCommerce Bookings: by WooCommerce – 1.15.68
WooCommerce Google Analytics Integration: by WooCommerce – 1.5.17

### Settings ###

API Enabled: –
Force SSL: –
Currency: EUR (€)
Currency Position: left
Thousand Separator: ,
Decimal Separator: .
Number of Decimals: 2
Taxonomies: Product Types: booking (booking)
external (external)
grouped (grouped)
simple (simple)
variable (variable)

Taxonomies: Product Visibility: exclude-from-catalog (exclude-from-catalog)
exclude-from-search (exclude-from-search)
featured (featured)
outofstock (outofstock)
rated-1 (rated-1)
rated-2 (rated-2)
rated-3 (rated-3)
rated-4 (rated-4)
rated-5 (rated-5)

Connected to WooCommerce.com: –
Enforce Approved Product Download Directories: ✔

### WC Pages ###

Shop base: #5 - /shop/
Cart: #6 - /cart/
Checkout: #7 - /checkout/
My account: #8 - /my-account/
Terms and conditions: ❌ Page not set

### Theme ###

Name: Twenty Twenty-Three
Version: 1.0
Author URL: https://wordpress.org
Child Theme: ❌ – If you are modifying WooCommerce on a parent theme that you did not build personally we recommend using a child theme. See: How to create a child theme
WooCommerce Support: ✔

### Templates ###

Overrides: –

### Admin ###

Enabled Features: activity-panels
analytics
coupons
customer-effort-score-tracks
experimental-products-task
experimental-import-products-task
experimental-fashion-sample-products
shipping-smart-defaults
shipping-setting-tour
homescreen
marketing
multichannel-marketing
mobile-app-banner
navigation
onboarding
onboarding-tasks
remote-inbox-notifications
remote-free-extensions
payment-gateway-suggestions
shipping-label-banner
subscriptions
store-alerts
transient-notices
woo-mobile-welcome
wc-pay-promotion
wc-pay-welcome-page

Disabled Features: minified-js
new-product-management-experience
settings

Daily Cron: ✔ Next scheduled: 2022-11-29 18:06:00 +00:00
Options: ✔
Notes: 43
Onboarding: -

### Action Scheduler ###

Complete: 40
Oldest: 2022-11-25 18:17:56 +0000
Newest: 2022-11-28 23:59:21 +0000

Pending: 2
Oldest: 2022-11-29 22:53:37 +0000
Newest: 2022-12-25 23:12:57 +0000


### Status report information ###

Generated at: 2022-11-28 23:59:30 +00:00
`
```

</details>

**Comment by ricardo** [comment]
I was able to reproduce the error only after adding a new webhook endpoint from the [Stripe dashboard](https://dashboard.stripe.com/webhooks) using the latest API key. I was able to reproduce the issue both with Bancontact and a 3DS card.

<img width="300" alt="Stripe new webhook endpoint API version" src="https://user-images.githubusercontent.com/5509901/205378146-a439cfe1-e0e0-494e-8c7a-41f54dca5979.png">

I suspect this is related to Stripe having removed the `charges` property from the PaymentIntent in the [latest API version `2022-11-15`](https://stripe.com/docs/upgrades#2022-11-15):

> The charges property on [PaymentIntent](https://stripe.com/docs/api/payment_intents/object) has been removed. You can use the latest_charge property instead.

### How to fix

As a temporary fix, you can create the webhook endpoint with an API version older than the latest `2022-11-15`.

Added this ticket to high priority.

**Comment by maxlaf** [comment]
5759571-zen
Not using UPE. Same issue with 3DS cards. Changing the title to remove UPE based on Ricardo's findings and that ticket.

@ricardo I'm unable to select a previous version when creating a webhook endpoint if my account is already on the latest API version.

![dha8nS.png](https://user-images.githubusercontent.com/56649567/205506168-5716c923-f3d5-4c0f-845b-0297781ed1ae.png)

From Stripe's docs, [Rolling back your API version](https://stripe.com/docs/upgrades#rolling-back-your-api-version), it seems you can only roll back to a previous version within 72 hours of the upgrade. We may have merchants that upgrade their API and don't roll back within 72 hours, potentially leaving them without a workaround for now.

**Comment by dreamtooloud** [comment]
To add to @maxlaf 's comment, the user in 5759571-zen does not have the option to roll back to a previous version of the API, they sent us a screenshot showing that it is not available to them: 

![https://d.pr/i/cmHxYO](https://d.pr/i/cmHxYO+) 
Full Size: https://d.pr/i/cmHxYO

So it does seem as though there is no work around right now (unless there is another spot they can look to roll back the API version that we haven't covered here).

**Comment by ricardo** [comment]
True. They can't roll back after 72 hours.

I guess the only immediate option would be to use a different Stripe account. I guess the 2022-11-15 version is not the default.

We have a fix in progress for the next release: https://github.com/woocommerce/woocommerce-gateway-stripe/pull/2493

**Comment by ryanr14** [comment]
Pretty sure 5750332-zen was affected by this. They have the latest API version as it's a new account. They wanted to be messaged again once a fix is released.

**Comment by maxlaf** [comment]
5768447-zen

**Comment by Brianmitchtay** [comment]
Seeing what appears to be this same issue in 5759571-zen as well

**Comment by dreamtooloud** [comment]
5790037-zen
5800282-zen
5791226-zen

**Comment by fetenlakhal** [comment]
33729708-hc

**Comment by maxlaf** [comment]
5755186-zen

**Comment by diegocurbelo** [comment]
We have a fix under review, and we plan to ship a patch release next week.

> Workaround
Disabling UPE and/or switching to a Stripe API version before 2022-11-15 solves these issues.

**Comment by dreamtooloud** [comment]
5810175-zen

**Comment by maxlaf** [comment]
5806592-zen

**Comment by marissaexplores** [comment]
[5834562-zen](5834562-zd-woothemes)

**Comment by maxlaf** [comment]
Probably 5837757-zen

**Comment by darcie** [comment]
5844007-zen
I've tested and as per the edit on the original post, disabling UPE does not fix this.

**Comment by drwpcom** [comment]
@diegocurbelo when was the fix released? There are lots of reports here after the issue is closed and it would be helpful to have a reference to the fix/release.

Related, are we using a 2019 version of the API as linked [here](https://woocommerce.com/document/stripe/#what-version-of-the-stripe-api-does-this-extension-use)?

**Comment by drwpcom** [comment]
For completeness, noting here that version 2022-11-15 of the Stripe API changed the structure of the webhook payload, we included a [fix for that](https://github.com/woocommerce/woocommerce-gateway-stripe/issues/2488) in version 7.0.2 of the Stripe plugin (released on Jan 11). H/T @diegocurbelo
